### PR TITLE
feat(core): validate custom client metadata when post or patch application

### PR DESCRIPTION
**Pull request by IceHe**
## Summary

Validate custom client metadata when POST /applications or PATCH /applications/:id

## Linear Issue Reference

LOG-2103
LOG-1885

## Testing

image

### CODE REVIEW

**Review by wangsijie — APPROVED** [review]

File: packages/core/src/oidc/utils.ts [hunk 1]
```diff
@@ -27,19 +27,21 @@ const isOrigin = (value: string) => {
   }
 };
 
+export const validateCorsAllowedOrigins = (value: unknown) => {
+  if (Array.isArray(value) && value.some((origin) => !isOrigin(origin))) {
+    throw new errors.InvalidClientMetadata(CustomClientMetadataKey.CorsAllowedOrigins);
+  }
+};
+
```
Comment (xiaoyijun) [inline packages/core/src/oidc/utils.ts:35]: We can use Zod for this verification?
Refer to https://github.com/logto-io/logto/pull/519/files

File: packages/core/src/oidc/utils.ts [hunk 2]
```diff
@@ -27,19 +27,21 @@ const isOrigin = (value: string) => {
   }
 };
 
+export const validateCorsAllowedOrigins = (value: unknown) => {
```
Comment (simeng-li) [inline packages/core/src/oidc/utils.ts:30]: should CORS all Uppercase?

File: packages/core/src/oidc/utils.ts [hunk 3]
```diff
@@ -27,19 +27,21 @@ const isOrigin = (value: string) => {
...
   }
 
-  if (
-    key === CustomClientMetadataKey.CorsAllowedOrigins &&
-    Array.isArray(value) &&
-    value.some((origin) => !isOrigin(origin))
-  ) {
-    throw new errors.InvalidClientMetadata(CustomClientMetadataKey.CorsAllowedOrigins);
+  if (key === CustomClientMetadataKey.CorsAllowedOrigins) {
+    validateCorsAllowedOrigins(value);
```
Comment (simeng-li) [inline packages/core/src/oidc/utils.ts:44]: The only reference of a util method is another util method. This implementation seems a little bit odd.

File: packages/core/src/oidc/utils.ts (hunk 1 above)
Comment (IceHe) [reply packages/core/src/oidc/utils.ts:35]: If use regex to check origin, it's too complex and not readable enough.

Comment (IceHe) [reply packages/core/src/oidc/utils.ts:35]: image

prefer to use `URL` to check origin

File: packages/core/src/oidc/utils.ts (hunk 2 above)
Comment (IceHe) [reply packages/core/src/oidc/utils.ts:30]: camelcase is acceptable.

image

See Notion [doc](https://www.notion.so/silverhand/Common-36449f7172514044967e5dbe537b87e9#3deb4db5eb7846b28ce2d61a376a0082)

File: packages/core/src/oidc/utils.ts (hunk 3 above)
Comment (IceHe) [reply packages/core/src/oidc/utils.ts:44]: 3 references.

another 2:

image
image

File: packages/core/src/oidc/utils.ts (hunk 2 above)
Comment (IceHe) [reply packages/core/src/oidc/utils.ts:30]: See coding [convention](https://www.notion.so/silverhand/Common-36449f7172514044967e5dbe537b87e9#3deb4db5eb7846b28ce2d61a376a0082)

image

File: packages/core/src/oidc/utils.ts (hunk 3 above)
Comment (IceHe) [reply packages/core/src/oidc/utils.ts:44]: Refactored. Implement it in another way. PTAL again.

File: packages/core/src/oidc/utils.ts (hunk 1 above)
Comment (IceHe) [reply packages/core/src/oidc/utils.ts]: Conclusion: use `zod.string().url().array()` to check

See Slack [thread](https://silverhand-io.slack.com/archives/C02A8G4HVAM/p1649821640292199).

**Review by xiaoyijun — APPROVED** [review]
LGTM

**Pull request merged**